infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisionerTest.java

```diff
@@ -0,0 +1,215 @@
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.testng.MockitoTestNGListener;
+import org.testng.Assert;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+/**
+ * Tests {@link ImagePullSecretProvisionner}.
+ *
+ * @author David Festal
+ */
+@Listeners(MockitoTestNGListener.class)
+public class ImagePullSecretProvisionerTest {
+
+  @Mock private KubernetesEnvironment k8sEnv;
+  @Mock private RuntimeIdentity runtimeIdentity;
+  @Mock private KubernetesClientFactory clientFactory;
+
+  @Mock
+  private UserSpecificDockerRegistryCredentialsProvider
+      userSpecificDockerRegistryCredentialsProvider;
+
+  @Mock private KubernetesNamespaceFactory namespaceFactory;
+  @Mock private KubernetesNamespace namespace;
+  @Mock private AuthConfigs authConfigs;
+  @Mock private KubernetesClient kubernetesClient;
+
+  @SuppressWarnings("rawtypes")
+  @Mock
+  private MixedOperation mixedOperation;
+
+  @SuppressWarnings("rawtypes")
+  @Mock
+  private NonNamespaceOperation namespaceOperation;
+
+  @Mock private Pod pod;
+  @Mock private PodSpec podSpec;
+  private LocalObjectReference existingImagePullSecretRef =
+      new LocalObjectReferenceBuilder().withName("existing").build();
+  private LocalObjectReference newImagePullSecretRef =
+      new LocalObjectReferenceBuilder().withName(ImagePullSecretProvisioner.SECRET_NAME).build();
+
+  private ImagePullSecretProvisioner imagePullSecretProvisioner;
+
+  @BeforeMethod
+  public void setup() throws InfrastructureException {
+    when(userSpecificDockerRegistryCredentialsProvider.getCredentials()).thenReturn(authConfigs);
+
+    when(k8sEnv.getPods()).thenReturn(ImmutableMap.of("wksp", pod));
+    when(pod.getSpec()).thenReturn(podSpec);
+    when(podSpec.getImagePullSecrets()).thenReturn(ImmutableList.of(existingImagePullSecretRef));
+
+    when(runtimeIdentity.getWorkspaceId()).thenReturn("workspaceId");
+
+    when(namespaceFactory.create(anyString())).thenReturn(namespace);
+    when(namespace.getName()).thenReturn("namespaceName");
+
+    when(clientFactory.create()).thenReturn(kubernetesClient);
+    when(clientFactory.create(anyString())).thenReturn(kubernetesClient);
+
+    doReturn(mixedOperation).when(kubernetesClient).secrets();
+    when(mixedOperation.inNamespace(anyString())).thenReturn(namespaceOperation);
+
+    imagePullSecretProvisioner =
+        new ImagePullSecretProvisioner(
+            userSpecificDockerRegistryCredentialsProvider, clientFactory, namespaceFactory);
+  }
+
+  @Test
+  public void dontDoAnythingIfNoPrivateRegistries() throws Exception {
+    when(authConfigs.getConfigs()).thenReturn(Collections.emptyMap());
+
+    imagePullSecretProvisioner.provision(k8sEnv, runtimeIdentity);
+
+    verifyZeroInteractions(namespaceOperation);
+    verifyZeroInteractions(podSpec);
+  }
+
+  private static class TestAuthConfig implements AuthConfig {
```

Comment:
Please, move auxiliary methods down. It is easier to read test when test cases are higher and all other methods are lower in the class.